classes/services/PKPSubmissionFileService.inc.php

```diff
@@ -0,0 +1,837 @@
+				$allowedFileStages[] = SUBMISSION_FILE_PROOF;
+			}
+			if ($hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_PRODUCTION_READY;
+			}
+		}
+
+		HookRegistry::call('SubmissionFile::assignedFileStages', [&$allowedFileStages, $stageAssignments, $action]);
+
+		return $allowedFileStages;
+	}
+
+	/**
+	 * Get all valid file stages
+	 *
+	 * @return array
+	 */
+	public function getFileStages() {
+		import('lib.pkp.classes.submission.SubmissionFile');
+		$stages = [
+			SUBMISSION_FILE_SUBMISSION,
+			SUBMISSION_FILE_NOTE,
+			SUBMISSION_FILE_REVIEW_FILE,
+			SUBMISSION_FILE_REVIEW_ATTACHMENT,
+			SUBMISSION_FILE_FINAL,
+			SUBMISSION_FILE_COPYEDIT,
+			SUBMISSION_FILE_PROOF,
+			SUBMISSION_FILE_PRODUCTION_READY,
+			SUBMISSION_FILE_ATTACHMENT,
+			SUBMISSION_FILE_REVIEW_REVISION,
+			SUBMISSION_FILE_DEPENDENT,
+			SUBMISSION_FILE_QUERY,
+		];
+
+		HookRegistry::call('SubmissionFile::fileStages', [&$stages]);
+
+		return $stages;
+	}
+
+	/**
+	 * Get the path to a submission's file directory
+	 *
+	 * This returns the relative path from the files_dir set in the config.
+	 *
+	 * @param int $contextId
+	 * @param int $submissionId
+	 * @return string
+	 */
+	public function getSubmissionDir($contextId, $submissionId) {
+		$dirNames = Application::getFileDirectories();
+		return sprintf(
+			'%s/%d/%s/%d',
+			str_replace('/', '', $dirNames['context']),
+			$contextId,
+			str_replace('/', '', $dirNames['submission']),
+			$submissionId
+		);
+	}
+
+	/**
+	 * Get the workflow stage for a submission file
+	 *
+	 * @param SubmissionFile $submissionFile
+	 * @return int|null WORKFLOW_STAGE_ID_*
+	 */
+	public function getWorkflowStageId($submissionFile) {
+		switch ($submissionFile->getData('fileStage')) {
+			case SUBMISSION_FILE_SUBMISSION:
+				return WORKFLOW_STAGE_ID_SUBMISSION;
+			case SUBMISSION_FILE_FINAL:
+			case SUBMISSION_FILE_COPYEDIT:
+				return WORKFLOW_STAGE_ID_EDITING;
+			case SUBMISSION_FILE_PROOF:
+			case SUBMISSION_FILE_PRODUCTION_READY:
+				return WORKFLOW_STAGE_ID_PRODUCTION;
+			case SUBMISSION_FILE_DEPENDENT:
+				$parentFile = Services::get('submissionFile')->get($submissionFile->getData('assocId'));
+				return $this->getWorkflowStageId($parentFile);
+			case SUBMISSION_FILE_REVIEW_FILE:
+			case SUBMISSION_FILE_REVIEW_ATTACHMENT:
+			case SUBMISSION_FILE_REVIEW_REVISION:
+				$reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO'); /* @var $reviewRoundDao ReviewRoundDAO */
+				$reviewRound = $reviewRoundDao->getBySubmissionFileId($submissionFile->getId());
+				return $reviewRound->getStageId();
+			case SUBMISSION_FILE_QUERY:
+				$noteDao = DAORegistry::getDAO('NoteDAO'); /* @var $noteDao NoteDAO */
+				$note = $noteDao->getById($submissionFile->getData('assocId'));
+				$queryDao = DAORegistry::getDAO('QueryDAO'); /* @var $queryDao QueryDAO */
+				$query = $queryDao->getById($note->getAssocId());
+				return $query ? $query->getStageId() : null;
+		}
+		return null;
```

Comment:
I'm tempted to suggest an exception rather than a `null` return. This case shouldn't ever happen in practice, I don't think, and null returns sneak through testing and are tougher to debug downstream.